parsl/executors/flux/executor.py

```diff
@@ -0,0 +1,247 @@
+from parsl.serialize import pack_apply_message, deserialize
+from parsl.providers import LocalProvider
+from parsl.app.errors import AppException
+
+
+class FluxFutureWrapper(cf.Future):
+    """Wrapper class around a ``flux.job.FluxExecutorFuture.``
+
+    Forwards methods onto the underlying FluxExecutorFuture.
+    """
+
+    def __init__(self, flux_future, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._flux_future = flux_future
+
+    def cancel(self):
+        if self._flux_future.cancelled():
+            # due to a bug, flux futures can only have ``cancel`` called once
+            return True
+        if self._flux_future.cancel():  # cancel underlying future and then self
+            if not super().cancel():
+                raise RuntimeError("Unexpected state")
+            self.set_running_or_notify_cancel()  # also should be called only once
+            return True
+        return False
+
+    cancel.__doc__ = cf.Future.cancel.__doc__
+
+    def running(self):
+        return self._flux_future.running()
+
+    running.__doc__ = cf.Future.running.__doc__
+
+
+def _complete_future(
+    expected_file: str, future_wrapper: FluxFutureWrapper, flux_future: Any
+):
+    """Callback triggered when a FluxExecutorFuture completes.
+
+    When the FluxExecutorFuture completes, check for the Parsl task's
+    output file, and assign the result to the FluxWrapperFuture future.
+
+    Parameters
+    ----------
+    expected_file : str
+        The path to the Parsl task's output file, storing the result of the task.
+    future_wrapper : FluxFutureWrapper
+        The user-facing future.
+    flux_future : FluxExecutorFuture
+        The future wrapped by ``future_wrapper``. Also accessible via
+        ``future_wrapper``, but the flux_future must be accepted as an argument
+        due to how ``concurrent.futures.add_done_callback`` works.
+    """
+    if flux_future.cancelled():  # if underlying future was cancelled, return
+        return  # no need to set a result on the wrapper future
+    try:
+        returncode = flux_future.result()
+    except Exception as unknown_err:
+        future_wrapper.set_exception(unknown_err)
+        return
+    if returncode == 0:
+        try:  # look for the output file
+            with open(expected_file, "rb") as file_handle:
+                future_wrapper.set_result(deserialize(file_handle.read()))
+        except FileNotFoundError:
+            future_wrapper.set_exception(
+                FileNotFoundError(
+                    f"No result found for Parsl task, expected {expected_file}"
+                )
+            )
+        except Exception as unknown_err:
+            future_wrapper.set_exception(unknown_err)
+    else:  # the job exited abnormally
+        future_wrapper.set_exception(
+            AppException(f"Parsl task exited abnormally: returned {returncode}")
+        )
+
+
+class FluxExecutor(NoStatusHandlingExecutor, RepresentationMixin):
```

Comment:
Currently the class inherits from NoStatusHandlingExecutor for convenience only. Flux provides status updates on all submitted jobs, so it would be pretty trivial to provide that on this executor. However, I see that ``Executor.status()`` is supposed to return a mapping from strings to ``JobStatus`` objects. What are the strings supposed to be? I already have a uuid for each Future returned from the executor, so it would make sense to use that string. However, how is the user supposed to figure out which uuid goes with which Future?

Reply:
Scaling and status handling in Parsl is fairly messy. The executor doesn't need to be a StatusHandlingExecutor if we don't expect to plug into Parsl's mechanism to gauge available resource (nodes/compute) and the current task load and then try to request through the executor to get additional nodes. Ultimately the question is whether we want any dynamic resource provisioning and if so whether Flux or Parsl should be making the provisioning decisions.

Reply:
If we just want a fixed-sized set of resources (the equivalent of 1 batch job) and just schedule tasks within it, `NoStatusHandlingExecutor` is probably the right choice.

Reply:
Flux doesn't currently have any way to request additional nodes or resources, I believe, although I know that is an area of research for the Flux team. I think you would have to request more nodes from the local resource manager and then launch Flux in the new allocation.

I mentioned in another comment that the executor can connect to remote Flux instances, so it might not be too hard to have a executor which communicates with multiple allocations (and hence multiple Flux instances). But it would take some work to get the load-balancing right across allocations, unless Parsl already has code to handle that.

I think that functionality, though, like the functionality of having a hierarchy of Flux instances controlled by the executor, could be added later.

Reply:
I've done some work to pull out HighThroughputExecutors block provisioning code into a superclass so it can also be used by the WorkQueueExecutor. The abstraction there is roughly that that kind of executor wants to execute task by having blocks of workers allocated by a batch system, but the nature of those workers is executor specific (so in the HightThroughputExecutor case, it is our process_worker_pool.py script; in the case of workqueue, it is workqueue's own `workqueue_worker` command. That abstraction also deals with deciding (in a simple way) when more blocks should be requested.

This abstraction doesn't prevent a user from saying "don't allocate any blocks at all - I will do this myself", which makes sense if a workqueue user wants to use some other provisioning tools provided by workqueue.

It sounds like potentially that abstraction fits your flux implementation too - if so, I can push on getting this into `master` a bit harder.

Reply:
That sounds great, I think that would fit well with this executor. However, what about load balancing? When a new app is submitted to the executor, how would the executor decide what blocks to forward the app onto? Is there any kind of abstraction for that or would that be up to me to implement? That could be pretty hard if each app has a variable amount of resources (cores and gpus).

Reply:
> When a new app is submitted to the executor, how would the executor decide what blocks to forward the app onto?

In the HighThroughputExecutor/WorkQueue model, that is one of the primary purposes of the executor implementation - they have some component (different in each) that the workers all connect with.

Reply:
And the workers report to the executor how much they are being utilized, so the executor knows which worker to pass new apps to?

Reply:
there's usually some kind of communication like that but it's executor-specific exactly what is going on there.

Reply:
Right, that makes sense. I'm just not sure there is great Flux functionality for figuring out whether a Flux instance is being underutilized or not. Off the top of my head the only thing I can think of is going by the number/percentage of PENDING jobs. I will talk to the Flux team about this.

Reply:
For reference: https://github.com/flux-framework/flux-core/discussions/3656